xds/csds/csds_test.go

```diff
@@ -0,0 +1,672 @@
+		endpointAnys[i], _ = ptypes.MarshalAny(endpoints[i])
+	}
+}
+
+func TestCSDS(t *testing.T) {
+	const retryCount = 10
+
+	xdsC, mgmServer, nodeID, stream, cleanup := commonSetup(t)
+	defer cleanup()
+
+	for _, target := range ldsTargets {
+		xdsC.WatchListener(target, func(client.ListenerUpdate, error) {})
+	}
+	for _, target := range rdsTargets {
+		xdsC.WatchRouteConfig(target, func(client.RouteConfigUpdate, error) {})
+	}
+	for _, target := range cdsTargets {
+		xdsC.WatchCluster(target, func(client.ClusterUpdate, error) {})
+	}
+	for _, target := range edsTargets {
+		xdsC.WatchEndpoints(target, func(client.EndpointsUpdate, error) {})
+	}
+
+	for i := 0; i < retryCount; i++ {
+		err := checkForRequested(stream)
+		if err == nil {
+			break
+		}
+		if i == retryCount-1 {
+			t.Fatalf("%v", err)
+		}
+		time.Sleep(time.Millisecond * 100)
+	}
+
+	if err := mgmServer.Update(e2e.UpdateOptions{
+		NodeID:    nodeID,
+		Listeners: listeners,
+		Routes:    routes,
+		Clusters:  clusters,
+		Endpoints: endpoints,
+	}); err != nil {
+		t.Fatal(err)
+	}
+	for i := 0; i < retryCount; i++ {
+		err := checkForACKed(stream)
+		if err == nil {
+			break
+		}
+		if i == retryCount-1 {
+			t.Fatalf("%v", err)
+		}
+		time.Sleep(time.Millisecond * 100)
+	}
+
+	const nackResourceIdx = 0
+	if err := mgmServer.Update(e2e.UpdateOptions{
+		NodeID: nodeID,
+		Listeners: []*v3listenerpb.Listener{
+			{Name: ldsTargets[nackResourceIdx], ApiListener: &v3listenerpb.ApiListener{}}, // 0 will be nacked. 1 will stay the same.
+		},
+		Routes: []*v3routepb.RouteConfiguration{
+			{Name: rdsTargets[nackResourceIdx], VirtualHosts: []*v3routepb.VirtualHost{{
+				Routes: []*v3routepb.Route{{}},
+			}}},
+		},
+		Clusters: []*v3clusterpb.Cluster{
+			{Name: cdsTargets[nackResourceIdx], ClusterDiscoveryType: &v3clusterpb.Cluster_Type{Type: v3clusterpb.Cluster_STATIC}},
+		},
+		Endpoints: []*v3endpointpb.ClusterLoadAssignment{
+			{ClusterName: edsTargets[nackResourceIdx], Endpoints: []*v3endpointpb.LocalityLbEndpoints{{}}},
+		},
+		SkipValidation: true,
+	}); err != nil {
+		t.Fatal(err)
+	}
+	for i := 0; i < retryCount; i++ {
+		err := checkForNACKed(nackResourceIdx, stream)
+		if err == nil {
+			break
+		}
+		if i == retryCount-1 {
+			t.Fatalf("%v", err)
+		}
+		time.Sleep(time.Millisecond * 100)
+	}
+}
+
+func commonSetup(t *testing.T) (xdsClientInterface, *e2e.ManagementServer, string, v3statuspb.ClientStatusDiscoveryService_StreamClientStatusClient, func()) {
+	t.Helper()
+
+	// Spin up a xDS management server on a local port.
+	nodeID := uuid.New().String()
+	fs, err := e2e.StartManagementServer()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a bootstrap file in a temporary directory.
+	bootstrapCleanup, err := e2e.SetupBootstrapFile(e2e.BootstrapOptions{
+		Version:              e2e.TransportV3,
+		NodeID:               nodeID,
+		ServerURI:            fs.Address,
+		ServerResourceNameID: "grpc/server",
```

Comment:
We don't have an xds-enabled gRPC-server here, right. So, I think we can skip this field.